### PR TITLE
Agent will operate with its default configuration out of the box

### PR DESCRIPTION
**Pull request by aritosteles**
|Related issue|
|---|
|[285](https://github.com/wazuh/wazuh-agent/issues/285)|

## Description
This PR aims to allow the agent to use default settings in case some or all settings are not specified in the configuration file or if the configuration file does not exist.

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux
  - [x] Windows
  - [x] MAC OS X

### CODE REVIEW

File: src/cmake/config.cmake [hunk 1]
```diff
@@ -4,11 +4,59 @@ set(PROJECT_NAME "Wazuh Agent")
...
 set(VERSION "0.1")
 
+# Default Agent Server Url
+set(DEFAULT_SERVER_URL "https://localhost:27000" CACHE STRING "Default Agent Server Url")
+
+# Default Agent Registration Url
+set(DEFAULT_REGISTRATION_URL "https://localhost:55000" CACHE STRING "Default Agent Registration Url")
+
+# Default Agent retry interval
+set(DEFAULT_RETRY_INTERVAL 30 CACHE STRING "Default Agent retry interval")
```
Comment (jr0me) [inline src/cmake/config.cmake:14]: There are many variables in this file that are defined as `STRING`, is it possible to use `INTEGER` for them?

Also, since we are adding the description to each, the comments are redundant, can we delete them?

File: src/modules/inventory/src/inventory.cpp [hunk 2]
```diff
@@ -1,11 +1,13 @@
...
+
 #include <cjson/cJSON.h>
+#include <config.h>
 #include <defs.h>
 #include <logger.hpp>
-
-#include <inventory.hpp>
 #include <sysInfo.hpp>
 
+#include <iostream>
```
Comment (jr0me) [inline src/modules/inventory/src/inventory.cpp:9]: It seems like this header is not needed.

File: src/agent/src/agent_registration.cpp [hunk 3]
```diff
@@ -1,9 +1,11 @@
 #include <agent_registration.hpp>
 
 #include <boost/beast/http.hpp>
-#include <iostream>
+#include <config.h>
 #include <nlohmann/json.hpp>
 
+#include <iostream>
```
Comment (jr0me) [inline src/agent/src/agent_registration.cpp:7]: Is `iostream` needed?

File: src/agent/configuration_parser/tests/configuration_parser_test.cpp [hunk 4]
```diff
@@ -1,9 +1,11 @@
 #include <configuration_parser.hpp>
 
+#include <config.h>
 #include <gtest/gtest.h>
 
 #include <filesystem>
 #include <fstream>
+#include <iostream>
```
Comment (jr0me) [inline src/agent/configuration_parser/tests/configuration_parser_test.cpp:8]: Is it needed here as well?

**Review by jr0me — COMMENTED** [review]
I'm not sure about the choice of interface for the default value, we are passing it all the way down to the configuration parser only to get it back if it can't find the requested key.

There's a similar end user case that you could achieve by changing the returned type to an optional, see https://en.cppreference.com/w/cpp/utility/optional/value_or .

Instead of `T GetConfig(T defaultValue, Keys... keys) const` and passing the default value all the way down, you can do `std::optional<T> GetConfig(Keys... keys)` and when used `configurationParser.GetConfig<int>("table", "key").value_or(DEFAULT_VALUE_FOR_KEY)`.

File: src/modules/inventory/src/inventory.cpp (hunk 2 above)
Comment (aritosteles) [reply src/modules/inventory/src/inventory.cpp:9]: Removed

File: src/agent/configuration_parser/tests/configuration_parser_test.cpp (hunk 4 above)
Comment (aritosteles) [reply src/agent/configuration_parser/tests/configuration_parser_test.cpp:8]: Removed

File: src/cmake/config.cmake (hunk 1 above)
Comment (aritosteles) [reply src/cmake/config.cmake:14]: - Implemented the `value_or()` change.
- Got rid of redundant comments.
- As for the STRING variables, the  only types available for [CACHE](https://cmake.org/cmake/help/v3.0/command/set.html) variables are FILEPATH, PATH, STRING, BOOL and INTERNAL. I'm not sure how this reflects on c++ constants, which for some reason where defined as `constexpr auto`, I just followed suit when adding new ones.

File: src/agent/communicator/include/communicator.hpp [hunk 5]
```diff
@@ -33,7 +33,7 @@ namespace communicator
                      std::string uuid,
                      std::string key,
                      std::function<std::string()> getHeaderInfo,
-                     const std::function<std::string(std::string, std::string)>& getStringConfigValue);
+                     const std::function<std::string(std::string, std::string, std::string)>& getStringConfigValue);
```
Comment (TomasTurina) [inline src/agent/communicator/include/communicator.hpp:36]: Revert this change.

File: src/agent/configuration_parser/include/configuration_parser.hpp [hunk 6]
```diff
@@ -71,8 +71,8 @@ namespace configuration
             }
             catch (const std::exception& e)
             {
-                LogError("The requested value could not be obtained: {}.", e.what());
-                throw;
+                LogWarn("Requested setting not found. Default value used. {}", e.what());
```
Comment (TomasTurina) [inline src/agent/configuration_parser/include/configuration_parser.hpp:74]:
```suggestion
                LogWarn("Requested setting not found, default value used: {}.", e.what());
```

File: src/agent/configuration_parser/tests/configuration_parser_test.cpp [hunk 7]
```diff
@@ -89,7 +90,7 @@ TEST(ConfigurationParser, GetConfigString)
           string_conf: string
     )";
     const auto parserStr = std::make_unique<configuration::ConfigurationParser>(strConfig);
-    const auto ret = parserStr->GetConfig<std::string>("agent", "server_url");
+    const auto ret = parserStr->GetConfig<std::string>("agent", "server_url").value_or("192.168.0.11");
```
Comment (TomasTurina) [inline src/agent/configuration_parser/tests/configuration_parser_test.cpp:93]: For these tests it may be better to set a different default value than the expected so they validate that the config is read.

File: src/cmake/config.cmake [hunk 8]
```diff
@@ -4,11 +4,40 @@ set(PROJECT_NAME "Wazuh Agent")
...
+
+set(BUFFER_SIZE 4096 CACHE STRING "Default Logcollector reading buffer size")
+
+set(DEFAULT_FILE_WAIT 500 CACHE STRING "Default Logcollector file reading interval (ms)")
+
+set(DEFAULT_RELOAD_INTERVAL 60 CACHE STRING "Default Logcollector reload interval (sec)")
+
+set(DEFAULT_LOGCOLLECTOR_ENABLED true CACHE BOOL "Default Logcollector enabled")
+
+set(DEFAULT_LOCALFILES "/var/log/auth.log" CACHE PATH "Default localfiles")
```
Comment (TomasTurina) [inline src/cmake/config.cmake:21]: Maybe this doesn't make any sense to be configured this way. We should directly set the default value in `config.h` file.

File: src/common/config/include/config.h.in [hunk 9]
```diff
@@ -1,12 +1,38 @@
...
+    constexpr auto DEFAULT_REGISTRATION_URL = "@DEFAULT_REGISTRATION_URL@";
+    constexpr auto DEFAULT_RETRY_INTERVAL = @DEFAULT_RETRY_INTERVAL@;
+    }
 
-constexpr auto BUFFER_SIZE = @BUFFER_SIZE@;
-constexpr auto DEFAULT_FILE_WAIT = @DEFAULT_FILE_WAIT@;
-constexpr auto DEFAULT_RELOAD_INTERVAL = @DEFAULT_RELOAD_INTERVAL@;
+    namespace logcollector
+    {
+        constexpr auto BUFFER_SIZE = @BUFFER_SIZE@;
```
Comment (TomasTurina) [inline src/common/config/include/config.h.in:17]: Note: In the future we need to evaluate if this setting should also be included in the `wazuh-agent.yml` file.

File: src/cmake/config.cmake (hunk 8 above)
Comment (aritosteles) [reply src/cmake/config.cmake:21]: Done

**Review by jr0me — APPROVED** [review]
LGTM

**Review by TomasTurina — APPROVED** [review]
